chrome/content/zotero/xpcom/progressQueue.js

```diff
@@ -0,0 +1,281 @@
+/*
+    ***** BEGIN LICENSE BLOCK *****
+    
+    Copyright © 2018 Center for History and New Media
+                     George Mason University, Fairfax, Virginia, USA
+                     http://zotero.org
+    
+    This file is part of Zotero.
+    
+    Zotero is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+    
+    Zotero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+    
+    You should have received a copy of the GNU Affero General Public License
+    along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+    
+    ***** END LICENSE BLOCK *****
+*/
+
+Zotero.ProgressQueue = function (options) {
+	const OFFLINE_RECHECK_DELAY = 60 * 1000;
+	
+	let _id = options.id;
+	let _title = options.title;
+	let _columns = options.columns;
+	let _processor = options.processor;
+	
+	let _listeners = {};
+	let _rows = [];
+	let _queue = [];
+	let _queueProcessing = false;
+	
+	
+	this.getId = function () {
+		return _id;
+	};
+	
+	
+	this.getTitle = function () {
+		return _title;
+	};
+	
+	
+	this.getColumns = function () {
+		return _columns;
+	};
+	
+	
+	/**
+	 * Add listener
+	 * @param name Event name
+	 * @param callback
+	 */
+	this.addListener = function (name, callback) {
+		_listeners[name] = callback;
+	};
+	
+	
+	/**
+	 * Remove listener
+	 * @param name Event name
+	 */
+	this.removeListener = function (name) {
+		delete _listeners[name];
+	};
+	
+	
+	/**
+	 * Adds items to the queue and triggers processing
+	 * @param items {Zotero.Item}
+	 */
+	this.addItems = function (items) {
+		for (let item of items) {
+			_addItem(item);
+		}
+		_processQueue();
+	};
+	
+	
+	/**
+	 * Returns all rows
+	 * @return {Array}
+	 */
+	this.getRows = function () {
+		return _rows;
+	};
+	
+	
+	/**
+	 * Returns rows count
+	 * @return {Number}
+	 */
+	this.getTotal = function () {
+		return _rows.length;
+	};
+	
+	
+	/**
+	 * Returns processed rows count
+	 * @return {Number}
+	 */
+	this.getProcessedTotal = function () {
+		return _rows.filter(row => row.status > Zotero.ProgressQueue.ROW_PROCESSING).length;
+	};
+	
+	
+	/**
+	 * Stop processing items
+	 */
+	this.cancel = function () {
+		_queue = [];
+		_rows = [];
+		if (_listeners['empty']) {
+			_listeners['empty']();
+		}
+	};
+	
+	
+	/**
+	 * Add item for processing
+	 * @param item
+	 * @return {null}
+	 */
+	function _addItem(item) {
+		for (let row of _rows) {
+			if (row.id === item.id) {
+				if (row.status > Zotero.ProgressQueue.ROW_PROCESSING) {
+					_deleteRow(row.id);
+					break;
+				}
+				return null;
```

Comment:
Just `return;`, unless there's a meaningful difference between `null` and `undefined`.

Reply:
Let's add a comment to `addItems()` clarifying this behavior — i.e., that if an item is already processed, it's reprocessed, and if it's in the queue but not done, it's ignored.